### ISSUE TITLE
Question about policy_loss

### ISSUE THREAD

**Issue by toshikwa** [pranz24/pytorch-soft-actor-critic#14]
Hi, thank you for your great work!!

I have a question related to #10.
Can you explain the meaning of the code below in the `GaussianPolicy`??

```
# Enforcing Action Bound
log_prob -= torch.log(1 - action.pow(2) + epsilon)
```
Also, can you provide some information which you referenced to code this loss??

Anyway, thank you for sharing your great codes.

**Comment by toshikwa** [comment]
I just realized the meaning and close it.

**Comment by Wen-Wen-Luffy** [comment]
Can you explain the line? Thank you very much!

**# Enforcing Action Bound
log_prob -= torch.log(self.action_scale * (1 - y_t.pow(2)) + epsilon)**

**Comment by toshikwa** [comment]
In gaussian policies, action is calculated like this.

1. x ~ Gaussian(mean, std)
2. y = tanh(x)
3. a = k * y

Here is how to calculate log likelihood of action (to get entropy).

1. likelihood of x is
p(x) = 1 / (sqrt(2*pi) * std) * exp(-1/(2*std^2) * (x - mean)^2)

2. likelihood of y is
p(y) = p(x) / |dy/dx| = p(x) / (1 - tanh(x)^2) = p(x) / (1 - y^2)

3. likelihood of a is
p(a) = p(y) * |dy/da| = p(y) / k = p(x) / (k * (1 - y^2))

Finally, you get this.
log(p(a)) = log(p(x)) - log(k*(1 - y^2))

**Comment by Wen-Wen-Luffy** [comment]
**Thank you very much!**

It's very clear.

**Comment by sunnyswag** [comment]
@ku2482 Thanks for your explanation.
However, I can't understand why we need to minus the log(p(a)).
Can you do a further explanation? I havn't found any description in the paper of SAC . Perhaps, I have overlook sth.

**Comment by toshikwa** [comment]
Hi, @sunnyswag

I found some typo in my explanation above, and fixed them.
I think they made it difficult to understand, so please read it again.

Also, let me assure you that we calculate log probability of action, log p(a), like below.

1. first calculate log probability of x, log p(x).

log_prob = ...

2. then, minus log(k*(1 - y^2)) to get log probability of a, log p(a).

log_prob -= torch.log(self.action_scale * (1 - y_t.pow(2)) + epsilon)


If you have somthing unclear, feel free to ask me.

**Comment by sunnyswag** [comment]
Hi, @ku2482 
Thank you . I have understood how we calculate ```log(p(a))``` thoroughly, i think.
However, the question this time I have meet is that why we need to do this operation :
```log_prob -=  log(p(a))```
In general, We just need ```log_prob``` if we need to calculate the entropy. For example,  PPO and A2C in continuous actionspace control.
I think the most likely thing is that we need to calculate KL-devergence after I browse the sac paper again.(Perhaps, I was wrong.) But I can't find the best matched explanation.

**Comment by toshikwa** [comment]
@sunnyswag

I'm sorry but I don't remember details of codes.
Can u tell me where the operation "log_prob -= log(p(a)) " is done??

I'm not sure I do that operation.

**Comment by sunnyswag** [comment]
Thanks for your reply. @ku2482 

u are right. I think I have figout the problem finally, but I'm not sure.
I can understand the formula you have written.

But my question is that why we need do such operation, why don't we just use ```log_porb = normal.log_prob(action)```, and then ```return log_prob , ...```.

In my opinion, since we use ```torch.tanh(action)``` to get the action, so we need to keep it consistency. And then we calculate the derivative of ```tanh(y)``` to get the real log_prob, is it right?

**Comment by toshikwa** [comment]
@sunnyswag

Due to tanh() operation, which is non linear, probability density distribution of the action, p(a), is not Normal distribution.

So we can't calculate log probability of action in the way you mentioned.

> And then we calculate the derivative of tanh(y) to get the real log_prob, is it right?

yes.

To get log p(a), I think it's the simplest way to first calculate log p(x), then calculate log p(a).

**Comment by sunnyswag** [comment]
@ku2482 
Ok, happy ending. 😃 

Thank you !
